analytics/lateral-movement-detection/DetectLM.py

```diff
@@ -0,0 +1,271 @@
+    temp=dfdata[['source','destination','id']].groupby(['source','destination'])['id'].count().reset_index().rename(columns={"id":"Count"})
+    temp=temp.groupby(['destination'])['Count'].max().reset_index().rename(columns={"Count":"Max"})
+    destinations=pd.merge(destinations,temp, how='inner', on = ['destination'])
+       
+    mms = MinMaxScaler()
+    mms.fit(destinations[['F1','F2','NoOfRecords','Mean','Max','Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su', '0to6', '6to12', '12to18', '18to0']])
+    destinations[['F1','F2','NoOfRecords','Mean','Max','Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su', '0to6', '6to12', '12to18', '18to0']] = mms.transform(destinations[['F1','F2','NoOfRecords','Mean','Max','Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su', '0to6', '6to12', '12to18', '18to0']])
+    kmeans = KMeans(n_clusters=kd,random_state=0)
+    destinations['c_dst'] = kmeans.fit_predict( destinations[['F1','F2','NoOfRecords','Mean','Max','Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa', 'Su', '0to6', '6to12', '12to18', '18to0']])
+    
+    return   destinations
+    
+
+                                 
+if __name__ == "__main__":
+    out = pd.read_parquet(DATA_PATH)
+    in1 = pd.read_parquet(DATA_PATH_1)
+    in2 = pd.read_parquet(DATA_PATH_2)
+    out['index1'] = out.index
+    
+    #data=pd.merge(in1,in2, how='inner', on =['first_observed'])
+    data=pd.merge(in1,in2, left_index=True, right_index=True)
+    data['index1'] = data.index
+    data['first_observed']=data['first_observed_x']
+    data['id']=data['id_x']
+    data['source']=data['src_ref.value']
+    data['destination']=data['dst_ref.value']
+    data['username']=data['user_id']
+    #dfdata['domain']=dfdata['objects.3.value']
+
+    dftest=data[data['status'].astype(str).str.match('unknown')].copy()        
+    dftrain=data[data['status'].astype(str).str.match('benign')].copy()                                                      
+    dftest=process(dftest)
+    dftrain=process(dftrain)  
+    print(len(dftest.index))  
+    print(len(dftrain.index)) 
+    csources=sourceclustering(dftrain)[['source','c_src']]
+    dftrain=pd.merge(csources,dftrain, how='inner', on = ['source'])
+    cusernames=usernameclustering(dftrain)[['username','c_usr']]
+    dftrain=pd.merge(cusernames,dftrain, how='inner', on = ['username'])
+    cdestination=destinationclustering(dftrain)[['destination','c_dst']]
+    dftrain=pd.merge(cdestination,dftrain, how='inner', on = ['destination'])
+    
+    dftest=pd.merge(csources[['source','c_src']],dftest, how='right', on = ['source'])
+    dftest=pd.merge(cusernames[['username','c_usr']],dftest, how='right', on = ['username'])
+    dftest=pd.merge(cdestination[['destination','c_dst']],dftest, how='right', on = ['destination'])
+    dftest=pd.merge(dftrain[['c_usr','c_src','c_dst','status']].drop_duplicates(),dftest[['index1','first_observed','source','destination','username','c_usr','c_src','c_dst']], how='right', on = ['c_usr','c_src','c_dst'])
```

Comment:
What 's happening with the indentation here?

Reply:
I used beautifier in new commit.